bot/exts/utilities/challenges.py

```diff
@@ -0,0 +1,309 @@
+            other_info_embed: Embed,
+            main_embed: Embed,
+    ):
+        options = [
+            SelectOption(
+                label="Main Information",
+                description="See the kata's difficulty, description, etc.",
+                emoji='🌎',
+            ),
+            SelectOption(
+                label='Languages',
+                description='See what languages this kata supports!',
+                emoji=Emojis.reddit_post_text,
+            ),
+            SelectOption(
+                label='Tags',
+                description='See what categories this kata falls under!',
+                emoji=Emojis.stackoverflow_tag,
+            ),
+            SelectOption(
+                label='Other Information',
+                description='See how other people performed on this kata and more!',
+                emoji='ℹ',
+            ),
+        ]
+
+        self.original_message = original_message
+        # maps embeds to item chosen
+        self.mapping_of_embeds = {
+            "Main Information": main_embed,
+            "Languages": language_embed,
+            "Tags": tags_embed,
+            "Other Information": other_info_embed,
+        }
+
+        super().__init__(
+            placeholder='See more information regarding this kata',
+            min_values=1,
+            max_values=1,
+            options=options,
+        )
+
+    async def callback(self, interaction: Interaction) -> None:
+        """Callback for when someone clicks on a dropdown."""
+        # Edit the message to the embed selected in the option
+        result_embed = self.mapping_of_embeds[self.values[0]]
+        await self.original_message.edit(embed=result_embed)
+
+
+class Challenges(commands.Cog):
+    """
+    Cog for a challenge command.
+
+    The challenge command pulls a random kata from codewars.com.
+    A kata is the name for a challenge, specific to `codewars.com`.
+
+    The challenge command also has filters to customize the kata that is given.
+    You can specify the language the kata should be from, the difficulty of the kata.
+    Lastly, you can customize the topic you want the kata to be about!
+    """
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    async def kata_id(self, search_link: str, params: dict) -> str:
+        """
+        Using bs4, and with a formatted link with the filters provided by the user.
+
+        This will webscrape the search page with `search_link` and then get the ID of a kata for the
+        codewars.com API to use.
+        """
+        async with self.bot.http_session.get(search_link, params=params) as response:
+            if response.status != 200:
+                logger.error(
+                    f"Unexpected status code {response.status} from codewars.com"
+                )
+            soup = BeautifulSoup(await response.text(), features="lxml")
+            first_kata_div = soup.find_all("div", class_="item-title px-0")
```

Comment:
The Python discord bot implements bs4 not in another thread iirc, and it's been working fine, but this is a valid point

Reply:
There is the option to run the calculation in a process pool (relevant docs [here](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor) and [here](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor)) if this does take up a lot of processing, this would allow the main event loop be freed to do other things.

Reply:
Thank you so much! So essentially, asyncio can run blocking code in a ThreadPoolExecutor (threading) and the main event loop can do other things?

Reply:
I'd rather use a the process pool executer (the second here) since this uses the multiprocessing feature allowing us to bypass the GIL

Reply:
What would be the use of bypassing the GIL here? The threads can still switch and I don't think we need to utilize more of the CPU.

From some tests I've done for the docs bs4 took quite some time for longer HTML so if it blocks for more than a ms or so I'd say it's definitely worth it, spawning threads is cheap

Reply:
Using threading could still potentially block the event loop, which is why I suggested using Multiprocessing, but yeah it is a bit unnecessary.

Reply:
No it won't potentially do so. Threading will not block the event loop.

The thing with threading is that it won't go faster, because it is not run in parallel and even when it is ran concurrently Python has the GIL which means that even concurrently only one thread can run at a time.. so half concurrently!

But that doesn't matter and we don't care, this is not a high priority calculation that we need to speed up with parallelism. The objective is to just not make it not block the event loop.